examples/vision/fully_convolutional_network.py

```diff
@@ -0,0 +1,651 @@
+  - FCN-32S
+  - FCN-16S
+  - FCN-8S
+- Load weights into converted Convolution layers
+- Training
+  - FCN-32S
+  - FCN-16S
+  - FCN-8S
+- Visualizations
+  - Metrics
+  - Segmentation Mask results
+
+To get a better idea on Image Segmentation or find more pre-trained models, feel free to
+navigate to the links below
+- [Hugging Face Image
+Segmentation](https://huggingface.co/models?pipeline_tag=image-segmentation)
+- [PyImageSearch Blog on Semantic
+Segmentation](https://pyimagesearch.com/2018/09/03/semantic-segmentation-with-opencv-and-deep-learning/)
+
+"""
+
+"""
+## Setup Imports
+
+We use the imports for the following reasons
+- `tensorflow_datasets`: Utility to download our dataset
+- `matplotlib.pyplot`: Utility for plotting and visualizations
+- `numpy`: Utility for performing array manipulation operations
+
+We then set the Random Seed to make sure the results are reproducible over separate
+iterations. We also use `tf.data.AUTOTUNE` to set some parameters based on the current
+system's configuration.
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+import matplotlib.pyplot as plt
+import tensorflow_datasets as tfds
+import numpy as np
+
+keras.utils.set_random_seed(42)
+
+AUTOTUNE = tf.data.AUTOTUNE
+
+"""
+## Set configurations for notebook variables
+
+We set the required parameters for the experiment. The chosen dataset has a total of 4
+classes per image, with regards to the segmentation mask. We also set our hyperparameters
+in this cell.
+
+Mixed Precision as an option is also available in systems which support it, to reduce
+load. This would make most tensors use `16-bit float` values instead of `32-bit float`
+values, in places where it will not adversely affect computation.
+"""
+
+config = {
+    "num_classes": 4,
+    "input_height": 224,
+    "input_width": 224,
+    "learning_rate": 1e-3,
+    "weight_decay": 1e-4,
+    "epochs": 10,
+    "batch_size": 32,
+    "mixed_precision": False,
+    "shuffle": True,
+}
+
+# Mixed-precision setting
+if config["mixed_precision"]:
+    policy = keras.mixed_precision.Policy("mixed_float16")
+    keras.mixed_precision.set_global_policy(policy)
+
+NUM_CLASSES = config["num_classes"]
+
+"""
+## Load dataset
+
+We make use of the [Oxford-IIIT Pets dataset](http://www.robots.ox.ac.uk/~vgg/data/pets/)
+which contains a total of 7,349 samples and their segmentation masks. The dataset is
+lightly imbalanced, but not enough to create a severe impact. We have 37 classes, with
+almost 200 samples per class.
+Our training and validation dataset has 3,128 and 552 samples respectively. Aside from
+this, our test split has a total of 3,669 samples.
+
+We set a `batch_size` parameter that will batch our samples together, use a `shuffle`
+parameter to mix our samples together.
+"""
+
+(train_ds, valid_ds, test_ds) = tfds.load(
+    "oxford_iiit_pet",
+    split=["train[:85%]", "train[85%:]", "test"],
+    batch_size=config["batch_size"],
+    shuffle_files=config["shuffle"],
```

Comment:
We may need to reshuffle every epoch using `train_ds = train_ds.shuffle()`?

Reply:
We can do that, yeah. Will add it in place.